### ISSUE TITLE
Glob sometimes hangs

### ISSUE THREAD

**Issue by ohkinozomu** [bmatcuk/doublestar#72]
Hello.
My use of this library is to extract files in a particular programming language from a git repository.

I have found that the program hangs in some repositories.

The following is sample code.

```go
package main

import (
	"fmt"
	"os"

	"github.com/bmatcuk/doublestar/v4"
)

func main() {
	pattern := "**/*.go"
	path := "../k3os"
	fsys := os.DirFS(path)
	files, err := doublestar.Glob(fsys, pattern)
	if err != nil {
		panic(err)
	}
	fmt.Println(files)
}
```

The two I have found are [rancher/k3os](https://github.com/rancher/k3os) and [google/google-ctf](https://github.com/google/google-ctf).

doublestar version: 4.2.0

**Comment by bmatcuk** [comment]
These repos have circular symbolic links. For example, in k3os, `images/30-bin/root` is a symbolic link to `../..` (ie, the root of the repo). So Glob is recursing on these endlessly. I'll have to think about the best way to solve that...

**Comment by bmatcuk** [comment]
@ohkinozomu I've [opened a new PR](https://github.com/bmatcuk/doublestar/pull/82) that adds a `WithNoFollow` option that can address this problem. You'd use it such as: `doublestar.Glob(fsys, pattern, doublestar.WithNoFollow())`. It should prevent the issue you were seeing with circular symlinks. Could you give it a try when you have a chance?

**Comment by ohkinozomu** [comment]
@bmatcuk
Thank you, the following code works.

```go
package main

import (
	"fmt"
	"os"

	"github.com/bmatcuk/doublestar/v4"
)

func main() {
	pattern := "**/*.go"
	path := "../k3os"
	fsys := os.DirFS(path)
	files, err := doublestar.Glob(fsys, pattern, doublestar.WithNoFollow())
	if err != nil {
		panic(err)
	}
	fmt.Println(files)
}
```

doublestar version: `v4.5.1-0.20230104180139-c5b3a48904d6`

**Comment by bmatcuk** [comment]
Awesome. I've merged the PR and released v4.6.0
